### ISSUE TITLE
Type 'X<false>' is not assignable to type 'X<boolean>'.

### ISSUE THREAD

**Issue by tomhanax** [microsoft/TypeScript#48150]
# Bug Report

Since 4.6 TS is complaining about that `false` is not assignable to `boolean`, when it is "wrapped by interface".

### 🔎 Search Terms

4.6 type

### 🕗 Version & Regression Information

- This is a compile problem
- This changed between versions 4.5.5 and 4.6

### ⏯ Playground Link

[Playground Link](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgPICMDO0Bud0A2EAPACoB8AUAN6XLIAUAlAFzKkDcdjeBArhDalWyOCACeXAL6VKMPiARhgAexDIVWXPiJlyDXgKEiM2KHkIkKNblAhg+UdQoAmEGKAgvRmURI70APSByGAAFijAALYADkRREOBwymrIwL4gKmBpsSpQYGJglDKUYOIxKAAKcOIJ4ACy9mEq3gC8yADkCHCYYR3IAD6d3VAuHQGyCGqY2QAebKbalsToKipEYuTI7ZpmFkQM8ATYTFxTIDPI4gta5jok1bWJYI3hLVs7t-sQDF09fadkJRzpcAF43Pb3YggPhRdDQADaAF0PhovvcGMjAUA)

### 💻 Code

```ts
interface Observable<T>
{
  (): T;
  (value: T): any;
}

function observable<T>(value: T): Observable<T>
{
  return undefined as any;  // the implementation is not important
}

const x: Observable<boolean> = observable(false);
```

### 🙁 Actual behavior

```ts
Type 'Observable<false>' is not assignable to type 'Observable<boolean>'.
  Types of parameters 'value' and 'value' are incompatible.
    Type 'boolean' is not assignable to type 'false'.
```

### 🙂 Expected behavior

No compiler problem as in previous versions.

**Comment by whzx5byb** [comment]
I believe this is a correct error which was not handled properly in old versions. 
The generic parameter `T` is invariant as it is used in both a covariant position `() => T` and a contravariant position `(value: T) => any`.

**Comment by tomhanax** [comment]
@whzx5byb Suppose this is correct behavior, can you suggest how to continue to use the sketched concept? The interface in fact represents the core concept of Knockout.js library. 

To set value to `x` - use `x(val)`. 
To read value from `x` - use `x()`.

 If `x` is defined to operate on `boolean`, It is very illogical that I can not set `true` or `false` as `val`. Of course I can cast to `boolean`, but is it the correct way? And is it the only way?

**Comment by whzx5byb** [comment]
> If `x` is defined to operate on `boolean`, It is very illogical that I can not set `true` or `false` as `val`. Of course I can cast to `boolean`, but is it the correct way? And is it the only way?

A way to make your example work:
```ts
function observable<T>(value: T extends infer U ? U: never): Observable<T>
{
  return undefined as any;  // the implementation is not important
}

const x: Observable<boolean> = observable(false); // no error
```

**Comment by tomhanax** [comment]
@whzx5byb  Just perfect, thanks!
I do not know if it is appropriate here, but if you want, you could post this as answer on https://stackoverflow.com/questions/71319489/type-observablefalse-is-not-assignable-to-type-observableboolean . Or I could do it using the code you proposed if you want. Please give me the info, I will close this bug then.

**Comment by bennieswart** [comment]
@tomhanax perhaps this should be referred to the knockout project so that the necessary changes can be made there.

**Comment by tomhanax** [comment]
> @tomhanax perhaps this should be referred to the knockout project so that the necessary changes can be made there.
Done: https://github.com/knockout/knockout/issues/2589, hope that somebody there will pick that up, because the project seem to go to winter sleep for good.

**Comment by whzx5byb** [comment]
@tomhanax I don't care, you could post it yourself.

**Comment by whzx5byb** [comment]
@tomhanax

The answer in https://github.com/microsoft/TypeScript/issues/48150#issuecomment-1060539236 is not perfect, as it will produce `unknown` without a type annotation.
```ts
const bad = observable(false);
//     ^ unknown, which should be Observable<boolean>
```

The issue here is the inference `T extends infer U ? U : never` infers `U` from the LHS of assignment. But when there is none, the default constraint type `unknown` will be used.

Here is [an alternative workaround](https://www.typescriptlang.org/play?ts=4.6.2#code/JYOwLgpgTgZghgYwgAgPICMDO0Bud0A2EAPACoB8yA3gFDLIAUAlAFzKkDcdjeBArhDalWyOCACeXAL40AJhAQE4UFDD4gEYYAHsQybVlz4iZADTIAquQa8BQ5BAAekELMzJQMaMgBKyAPy+yGwWIuoA1iDaAO56Ti5u7AFohlB4hCRWwSnYacYkFFw0YOIADigACnDiALYQ4ACyEGAAFtqyyAC8yADkCHCYLT3IAD69-VCyPRzINDQIuphgyI5sGLnpJuja2kRilN0GG-kM8ATYTFwLIEvI4mupmyRVtfVgTa3tB-qPJ30DQ1E7hedUazTaskus2utwAXg9jhliCA+DV0NAANoAXW+RyMGQY2Khc3mi2WUAAjF0foiiKc4OcIJcaERyQAmal4vIE-6DHrMmiUhhgKACJjIAD0Ev04UFbIYFPFUuQCygKk0DjV2igNCAA) which will produce the same behavior without a type annotation, but a bit more verbose. I have to introduce two type parameter. `T` is for the refered type which is from LHS. `U` is for the "raw" widening type, which will be used only when `T` is unknown.
```ts
declare function observable<T, U>(value: T extends infer R ? R : U): unknown extends T ? Observable<U> : Observable<T>;
```

I'm not sure if there is a better solution. If you find some please let me know.

**Comment by maskmaster** [comment]
Cross-posting here, just for the record. Make sure that also both these lines are valid with any new proposal.
```typescript
const x = observable<false | undefined>(false);
const y = observable(false);
```

**Comment by tomhanax** [comment]
@mbest, the "most updated" playground contains many examples and seems OK, of course anybody can check and correct if needed, I am no expert in this.

[Playground Link](https://www.typescriptlang.org/play?ts=4.6.2#code/JYOwLgpgTgZghgYwgAgPICMDO0Bud0A2EAPACoB8yA3gFDLIAUAlAFzKkDcdjeBArhDalWyOCACeXAL40AJhAQE4UFDD4gEYYAHsQybVlz4iZADTIAqsgC8oieQa8BAfiHIIAD0ghZmZKBhoZAAlZGcQ5DYLEXUAaxBtAHc9T29fdjC0Qyg8QhILSjYMbBzjEgouGjBxAAcUAAU4cQBbCHAAWQgwAAttWRtkAHIEOExuweQAHyGRqFlBjmQaGgRdTDBkDyLs3JN0bW0iMUpbAxLdiAZ4AmwmLg8GdXkYUAhZO-oAek-kVagVTTuf7aKA0B5gKACO5gq5wG4QaEPACMH2+vxBAI20CgIOWqxA62Q4m25zKxEaLTaYE6PT6U2QTwgLxAbxO+h2ZQYw1G4w+NHEDAARLNZILoQLhTyxVwBYzmW9xQwUcg0X9MUCcaCVmsNgAvElGPLEEB8ZroaAAbQAumyzoaiAxrXzdY6rUwaC6LUjTAAmN0ex2CuCCt1fH5qhRY4FQJbagkbACOAztpSN+0OEDE9Llr1kDmhCcePiZuYLDAhUK4heutyrStR4YxkY1uLjhJjpw5eVh8OhUCLz1Li1VTcB2Nx-ZrCK4-Yr05o-ZRw8b-2b461RA2NWTXYd3LGg2hNSFIulNGPkrGZ4vhGMd+lYfRq805kwoCQ-g2wD86ygoAA5ue9Yqiu6rrssm6iDupLdtCcD1lw8E5iy7wcEAA)

I am not saying anything about what "should" be fixed, I am no expert at all. Simply trying to help to solve existing issue, because now we are forced to remain on TS 4.5 until this is solved.

**Comment by maskmaster** [comment]
@tomhanax why was the issue closed? Isn't this a valid bug in Typescript?
As @mbest pointed out:

```typescript
const w = observable(false); // Typescript identifies 'w' as Observable<boolean>
const x: Observable<boolean> = observable(false); // Type being Observable<false> now
```
Typescript seems to change it's mind of what type `observable(false)` actually is.

**Comment by tomhanax** [comment]
@maskmaster to be honest I am not sure. I only know that what worked previously does not work now. As @whzx5byb pointed out, the whole thing may be "reversed" - it was working because of less perfect previous TS version that allowed this and now it is not working because simply it should not work in the first place. I am not good at type theory at all, so anybody else have to say funded words.
Anyway, I am reopening this in hope for clarification.

**Comment by whzx5byb** [comment]
> Typescript seems to change it's mind of what type observable(false) actually is.

This should be related to the "type widening" mechiasm. For more details: https://github.com/microsoft/TypeScript/pull/10676, https://github.com/microsoft/TypeScript/pull/11126, https://github.com/microsoft/TypeScript/pull/24310.

And here `boolean` behaves a little bit different because it is actually implemented as a union equivalent to `true | false`.
```ts
declare function fn<T>(x: T): T[];

const b1 = fn(true); // T is boolean (widening)
const n1 = fn(1);  // T is number (widening)

const b2: boolean[] = fn(true); // T is true (non-widening)
const n2: number[] = fn(1); // T is number (widening)

const b3: (true | false)[] = fn(true); // T is true (non-widening)
const n3: (1 | 0)[] = fn(1); // T is 1 (non-widening)
```

**Comment by maskmaster** [comment]
The following is valid
```typescript
const w = observable(false);
const v: Observable<boolean> = w;
```
However, this is not valid
```typescript
const v: Observable<boolean> = observable(false);
```
To me, it makes no sense.

**Comment by RyanCavanaugh** [comment]
@maskmaster any type with a contravariant position has this problem

**Comment by RyanCavanaugh** [comment]
Bisected to #47341. @ahejlsberg this seems like an unintended consequence?

**Comment by ahejlsberg** [comment]
Before 4.6 we'd measure variance of `Observable<T>` wrong:

```ts
declare let xb: Observable<boolean>;
declare let xf: Observable<false>;

xb = xf;  // Error, but wasn't before 4.6
xf = xb;  // Error
```

The type parameter `T` is used in both input and output positions within `Observable<T>`, so clearly `Observable<T>` is invariant and our current behavior is correct. We got it wrong before because we'd cross-relate the signatures, which sometimes makes sense for different source and target types, but never makes sense when source and target are instantiations of the same type (which they are when we're measuring variance). Not only was it wrong to cross-relate them, it also generated a lot more work, which was the problem fixed in #47341.

**Comment by maskmaster** [comment]
@ahejlsberg I totally agree with the correctness of your example and that the new behavior as you describe is correct.

I think we are talking about two different things.
In this simple example
```typescript
const w = observable(false);
```
According to Typescript 4.6 'w' is an `Observable<boolean>`.
Still, it is not possible to do (exactly what typescript suggests)
```typescript
const w: Observable<boolean> = observable(false);
```
According to you, what is the correct type of 'w' in `const w = observable(false)`?

**Comment by ahejlsberg** [comment]
@maskmaster Hmm, yeah, that is a known design limitation, and I agree it's unfortunate. See my comment [here](https://github.com/microsoft/TypeScript/issues/46707#issuecomment-963469208). The workaround is to write:

```ts
const w: Observable<boolean> = observable(false as boolean);
```

**Comment by ahejlsberg** [comment]
@maskmaster Regarding the type of

```ts
const w = observable(false);
```

it is producing the intended `Observable<boolean>`. There is no contextual type for the `false` argument and therefore we widen the type to `boolean`. Just like we would widen `42` to `number`.

**Comment by tomhanax** [comment]
> ```ts
> const w: Observable<boolean> = observable(false as boolean);
> ```

Bluntly said this is absolutely silly. In year 2022 you can not possibly tell this to the "average programmer". 
Anyway, what is the problem with "more verbose" typings illustrated in the beforementioned [playground](https://www.typescriptlang.org/play?ts=4.6.2#code/JYOwLgpgTgZghgYwgAgPICMDO0Bud0A2EAPACoB8yA3gFDLIAUAlAFzKkDcdjeBArhDalWyOCACeXAL40AJhAQE4UFDD4gEYYAHsQybVlz4iZADTIAqsgC8oieQa8BAfiHIIAD0ghZmZKBhoZAAlZGcQ5DYLEXUAaxBtAHc9T29fdjC0Qyg8QhILSjYMbBzjEgouGjBxAAcUAAU4cQBbCHAAWQgwAAttWRtkAHIEOExuweQAHyGRqFlBjmQaGgRdTDBkDyLs3JN0bW0iMUpbAxLdiAZ4AmwmLg8GdXkYUAhZO-oAek-kVagVTTuf7aKA0B5gKACO5gq5wG4QaEPACMH2+vxBAI20CgIOWqxA62Q4m25zKxEaLTaYE6PT6U2QTwgLxAbxO+h2ZQYw1G4w+NHEDAARLNZILoQLhTyxVwBYzmW9xQwUcg0X9MUCcaCVmsNgAvElGPLEEB8ZroaAAbQAumyzoaiAxrXzdY6rUwaC6LUjTAAmN0ex2CuCCt1fH5qhRY4FQJbagkbACOAztpSN+0OEDE9Llr1kDmhCcePiZuYLDAhUK4heutyrStR4YxkY1uLjhJjpw5eVh8OhUCLz1Li1VTcB2Nx-ZrCK4-Yr05o-ZRw8b-2b461RA2NWTXYd3LGg2hNSFIulNGPkrGZ4vhGMd+lYfRq805kwoCQ-g2wD86ygoAA5ue9Yqiu6rrssm6iDupLdtCcD1lw8E5iy7wcEAA)? I am only asking because that "simply works", but to be true, I really do not understand the magic there.

**Comment by ahejlsberg** [comment]
> Bluntly said this is absolutely silly

Yeah, as I said, it's unfortunate. I'll open an issue on this specifically and we'll see if we can do better.

**Comment by maskmaster** [comment]
Thank you @ahejlsberg for the explanation. It makes sense if it a design limitation. I'll try to wrap my head around this later on so I understand exaxtly what happens.
Please, could you also link the other issue that you opened/will open?

**Comment by DevProLog** [comment]
Hi,

It seems this issue is not yet solved.

I use Knockout with TS and am currently stuck using TS version 4.5.
Using higher versions of TS (currently trying v4.8.4), I get unexpected errors.
Half of the errors are fixed by the more specific definition suggested by @tomhanax.

```typescript
//export function observable<T>(value: T): Observable<T>;
export function observable<T, U = any>(value?: T extends infer R ? R : U): unknown extends T ? Observable<U> : Observable<T>;
export function observable<T = any>(value: null): Observable<T | null>
```

E.g.: Without the fix, the following results in a TS compiler error:

```typescript
AttestType: Observable<EAttestType> = ko.observable(EAttestType.Consultatie);
```

And, in order to please the TS compiler, would need to be changed to this:
```typescript
AttestType: Observable<EAttestType> = ko.observable<EAttestType>(EAttestType.Consultatie);
```
or this:
```typescript
AttestType: Observable<EAttestType> = ko.observable(EAttestType.Consultatie as EAttestType);
```

However, we still encounter other TS compiler errors, specifically related to types inheriting abstract classes in observables in the following form:

```typescript
export class ModalExistingPatientsComponentParams extends BaseComponentParams {
    ComponentViewModel: Observable<ModalExistingPatientsComponentViewModel | null> = observable<ModalExistingPatientsComponentViewModel | null>(null);

    constructor() {
        super();
    }
}

Property 'ComponentViewModel' in type 'ModalExistingPatientsComponentParams' is not assignable to the same property in base type 'BaseComponentParams'.
  Type 'Observable<ModalExistingPatientsComponentViewModel | null>' is not assignable to type 'Observable<BaseComponentViewModel | null>'.
    Types of parameters 'value' and 'value' are incompatible.
      Type 'BaseComponentViewModel | null' is not assignable to type 'ModalExistingPatientsComponentViewModel | null'.
```

TS says that it cannot assign the base type to the derived type, which is correct.
But the question is, why is it trying to do this in the first place?
I think that the knockout TS definitions are not supporting working with derived types in observables.

Here is a simplified playground detailing the issue:
[Playground link](https://www.typescriptlang.org/play?ts=4.8.4#code/JYOwLgpgTgZghgYwgAgPICMDO0Bud0A2EAPACoB8yA3gFDLIAUAlAFzKkDcdjeBArhDalWyOCACeXAL40aAejnIAJhAQE4UFDD4gEYYAHsQyA1lz4iZcg14ChIjNih5CJClxVqNWnXsPHTJxdLUmQAXlEJa1tBZBA+AgIHM2cLN2QAHziEgnIaT3VNZG1dfSMTFOC3ABpkAFVwyPFouH4IAH4hZAgAD0gQJUxkUBhoZAAlZHaJ5DY6kR0AaxADAHdjXv7B9im0SrTiOso2R3NXKy5ZXoAHAygwUSwwKEQHr0whgCE4bABhAwAtrcQBBwAA1YAQVYAWQMKgI1G49AAkpgADIGOBKUAAcxO+3O6AMBiIYkoEUCZyIDGeAiYXHoKMw-xAmFpeggSnxQQORJJEDJjUpqVcDHgBGw9NkjOQCCMbKgfD0dwY1w0cABmDY3z+gOBoLAAAV1ZqmIiZTK1S9NQA6f5AowGiFQ2HwmkAC2AmClMpkSOQiwMABEIJgkAMxGBmXqiJBGsxwpRaBbGWBPZgbaiWQqlZAlDTFRAfYyZH6bncHvgFa9ZeoPsgdRB7frwABlVRlEDOmFwiAI5OM1EYrG47lUkh80kgckVHmi2lFhmD6Os9l5sciyyTgXToUE6niyWXGVy1eK5VQVUmrUNn5NvWOtsd-zG63e80p5BWjUZ5uPsDtn4Rjdq6fYel6xb0H6MqBiGYagkokbRkCsYoBECZhEm-oymmXqZiuOYcvmC6QcgpZXD0tz3I81Z6LWPxfHef4guAr4-h+9BVs8NbMU6kI9vCG5VMQja8eC-GgQiWTxIk5DSBRVGVk8Lx0e8jG6g6LEAc+RhsZqHE0dxdFidpQFdhJvYEEJByiQ+WmAZ2IGWZk2SyUunHKTWr4GiZTmCbeGktmAfl9vJNDltRanICZLKQH0BnfpqACE2pMXZBp6UMWS2ZpBoOS+17Hoyp6EWAKrYYyiU3jlQWZS5NX-vlunXv6ZoDimuEZlVjRVe5ZE0GWlEVvR9auq0ACiPRevoIA4sa+gGshtXXt0fQIep965axK3tdF6XiS6lnWecY0EJN024vNkLgEt-4hVJrm5Huc6WKd51spdcALTdvkWfCLkybkDCA1K-olWuKptRV9CYHw1zQMwfXkeFQ2RXWQxvVNH2zVdi2-Yd-2bOtAWbUF90GUYrZwDgEC4+AvytAQ6CIIsKWMBhlA4AYwBKGa0k5H1ACCYCQGypDiPDx2WONwui2A4vwzOwpVAwMsi6G8sSxANqCwM0CLtKxXyhDl5VWwmMXTjX3XVGJmZVDn7ILD8Om9epHIxFDygnwALIGrcsKygu26yomjVP6qCzaGCCeosaYQDz4cymC0DXEQEA4osOjYjiEBJ4y2YJGA1sQANshAA)
